odata-protocol/11.4 Data Modification.md

```diff
@@ -655,49 +686,50 @@ operations.
+through a navigation property path binding [#OData-CSDL#NavigationPropertyPathBinding]
+or through the presence of a context URL within the payload;
+if the service is unable to determine it, it MUST fail the request.
 
 Upserts are not supported against entities whose keys' values are
-generated by the service. Services MUST fail an update request to a URL
+generated by the service. Services MUST fail a request to a URL
 that would identify such an entity and the entity does not yet exist.
 
 Similarly, services MUST fail an update request to the URL of a [media entity](#RequestingtheMediaStreamofaMediaEntityusingvalue) that does not yet exist.
-However, a `PUT` request to the _media edit URL_ of a media entity does have Upsert
+However, a `PUT` request to the _media edit URL_ of a media entity does have upsert
 semantics, in that the media entity is [created](#CreateaMediaEntity)
 with the specified media stream if it does not already exist, otherwise the
 media stream of the existing media entity is
 [updated](#UpdateaMediaEntityStream).
 
-Singleton entities can be upserted if they are nullable. Services
-supporting this SHOULD advertise it by annotating the singleton with the
-term `Capabilities.UpdateRestrictions` (nested property `Upsertable`
-with value `true`) defined in [OData-VocCap](#ODataVocCap).
-
 A key property whose value is provided in the request URL SHOULD be omitted from the request body.
 If key properties are provided in the request URL and the request body with different values,
 services MUST either fail the request or ignore the value in the request body.
 
-To ensure that an update request is not treated as an insert, the client
-MAY specify an [`If-Match`](#HeaderIfMatch) header in the update
-request. The service MUST NOT treat an update request containing an
-`If-Match` header as an insert.
-
-A `PUT` or `PATCH` request MUST NOT be treated as an update if an
-[`If-None-Match`](#HeaderIfNoneMatch) header is specified with a value
-of `*`.
+If the resource path is a canonical URL and the request
+- contains an [`If-Match: *`](#HeaderIfMatch) header, the service MUST fail the request
+  if it would be treated as an insert
+- contains an `If-Match` header with an ETag as value, the service treats the request as
+  an update if the [ETag](#UseofETagsforAvoidingUpdateConflicts) matches the addressed entity
+  or fails the request, but never treats it as an insert
+- contains an [`If-None-Match: *`](#HeaderIfNoneMatch) header, the service MUST fail the request
+  if it would be treated as an update.
+
+In other words, the `If-`(`None-`)`Match` header distinguishes between insert
+and update in this case. Otherwise the resource path ends with a
+non-containment navigation property and an `If-`(`None-`)`Match` header applies to
+the current related entity addressed by it (see [example ##antietag]).
```

Comment:
In the other case, the client may not care whether the operation is an update or upsert, and hence may not want to evaluate both `Updatable` and `Upsertable` in the capabilities annotations.